### PR TITLE
4758: Migrate project steps and groups

### PR DESCRIPTION
**Pull request by X0nic**
This will pull all project events into TimelineEntries. To keep the logic simple, it does this in more than one pass. Everything is pulled in excluding project parent ids, or schedule parent ids. Once those are inserted, the steps with children are converted to ProjectGroups. The children are queried for and then assigned parents. Then finally the schedule children are queried and given schedule parents. 

I have updated the migration script to logs errors to stderr instead of stdout. If you want a log with just errors, while still displaying all output to the screen to can run the migrations as such:

```sh
rake tww:migrate_all 2> migration_errors.log 2>&1
```

- [x] No date logic in ProjectGroup
- [x] Only migrate steps with Loan as ProjectTable
- [x] Ignore non-ProjectGroup columns when migrating steps with children (i.e. only copy the attribs that are specific to ProjectGroups.
- [x] If ProjectGroup has date dependency, copy to its children, but ONLY if they don’t have explicit start dates OR their own explicit dependencies.
- [x] Ensure Loan exists before assigning TimelineEntry to it. Ignore and report out any steps that have missing Loans.

### CODE REVIEW

**Comment by X0nic** [comment]
Error log:

```

❯ cat migration_errors.log
ProjectEvent[402] - mapping 0 MemberId ref to null
ProjectEvent[403] - mapping 0 MemberId ref to null
ProjectEvent[404] - mapping 0 MemberId ref to null
ProjectEvent[405] - mapping 0 MemberId ref to null
ProjectEvent[406] - mapping 0 MemberId ref to null
ProjectEvent[407] - mapping 0 MemberId ref to null
ProjectEvent[408] - mapping 0 MemberId ref to null
ProjectEvent[409] - mapping 0 MemberId ref to null
ProjectEvent[462] - mapping 0 MemberId ref to null
ProjectEvent[465] - mapping 0 MemberId ref to null
ProjectEvent[4139] - mapping 0 MemberId ref to null
ProjectEvent[14316] - Person not found for MemberId: 249, mapping MemberId ref to null
ProjectStep[16102] error migrating parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectStep[16103] error migrating parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectStep[16104] error migrating parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
ProjectStep[16105] error migrating parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
ProjectStep[16113] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16114] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16115] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16116] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16342] error migrating parent: Couldn't find TimelineEntry with 'id'=16336 - skipping
ProjectStep[14978] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3cfc0838> - skipping
ProjectStep[14986] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf37a993f0> - skipping
ProjectStep[14991] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c983998> - skipping
ProjectStep[14992] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c928e30> - skipping
ProjectStep[14995] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c828c88> - skipping
ProjectStep[14997] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a8f93a8> - skipping
ProjectStep[14998] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf392da310> - skipping
ProjectStep[15001] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf316ca158> - skipping
ProjectStep[15650] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a8a5dc0> - skipping
ProjectStep[15651] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a89f5d8> - skipping
ProjectStep[15652] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf315a4ee0> - skipping
ProjectStep[15653] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3156e660> - skipping
ProjectStep[15665] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a5bcaf0> - skipping
ProjectStep[15666] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a5661f0> - skipping
ProjectStep[15670] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a29bad8> - skipping
ProjectStep[15671] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a261428> - skipping
ProjectStep[15716] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf397f0368> - skipping
ProjectStep[15718] error migrating schedule parent: undefined method `scheduled_end_date' for #<ProjectGroup:0x007fcf31416150> - skipping
ProjectStep[15721] error migrating schedule parent: undefined method `scheduled_end_date' for #<ProjectGroup:0x007fcf396adc80> - skipping
ProjectStep[15728] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf39433450> - skipping
ProjectStep[15729] error migrating schedule parent: undefined method `scheduled_end_date' for #<ProjectGroup:0x007fcf393e2b18> - skipping
ProjectStep[15733] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf39257640> - skipping
ProjectStep[15734] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf391eccf0> - skipping
ProjectStep[15738] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf390b1980> - skipping
ProjectStep[15742] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf32507db0> - skipping
ProjectStep[15745] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf327bb638> - skipping
ProjectStep[15746] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf32759d70> - skipping
ProjectStep[15772] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf323e50e0> - skipping
ProjectStep[15773] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf30efcef8> - skipping
ProjectStep[15774] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf30e4d458> - skipping
ProjectStep[15788] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3ca43ec8> - skipping
ProjectStep[15793] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf31482aa8> - skipping
ProjectStep[15794] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3157caf8> - skipping
ProjectStep[15795] error migrating schedule parent: undefined method `scheduled_end_date' for #<ProjectGroup:0x007fcf31519598> - skipping
ProjectStep[15797] error migrating schedule parent: undefined method `scheduled_end_date' for #<ProjectGroup:0x007fcf3ca53e40> - skipping
ProjectStep[15802] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3cd58ed8> - skipping
ProjectStep[15803] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3ccda678> - skipping
ProjectStep[15804] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c94b8b8> - skipping
ProjectStep[15805] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3b029380> - skipping
ProjectStep[15806] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d092fe0> - skipping
ProjectStep[15807] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf397586d0> - skipping
ProjectStep[15808] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf37a6a6e0> - skipping
ProjectStep[15809] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3168ae68> - skipping
ProjectStep[15885] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d1502c0> - skipping
ProjectStep[15886] error migrating schedule parent: undefined method `scheduled_end_date' for #<ProjectGroup:0x007fcf3ccb0918> - skipping
ProjectStep[15893] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3cc7ac00> - skipping
ProjectStep[15897] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a8a3818> - skipping
ProjectStep[16062] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d1c3518> - skipping
ProjectStep[16063] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3ceb0b00> - skipping
ProjectStep[16064] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf392fa4a8> - skipping
ProjectStep[16065] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf315c7850> - skipping
ProjectStep[16066] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf37c311b8> - skipping
ProjectStep[16067] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf37b26570> - skipping
ProjectStep[16068] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3b0b7e00> - skipping
ProjectStep[16069] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d791468> - skipping
ProjectStep[16100] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d6c8bd0> - skipping
ProjectStep[16102] error migrating schedule parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectStep[16103] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d5f0168> - skipping
ProjectStep[16104] error migrating schedule parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
ProjectStep[16105] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a7a3828> - skipping
ProjectStep[16363] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d781838> - skipping
ProjectStep[16364] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a7ca4f0> - skipping
ProjectStep[16365] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d6a3998> - skipping
ProjectStep[16366] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3d660968> - skipping
ProjectStep[16379] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3cf612c0> - skipping
ProjectStep[16380] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3cf52630> - skipping
ProjectStep[16384] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3cd72590> - skipping
ProjectStep[16385] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3cd1b600> - skipping
ProjectStep[16386] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a974990> - skipping
ProjectStep[16395] error migrating schedule parent: undefined method `scheduled_end_date' for #<ProjectGroup:0x007fcf3a9476c0> - skipping
ProjectStep[16471] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c9a8748> - skipping
ProjectStep[16472] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c951e98> - skipping
ProjectStep[16473] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3b107018> - skipping
ProjectStep[16474] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c8d0438> - skipping
ProjectStep[16475] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c8a12c8> - skipping
ProjectStep[16476] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3c82a6a0> - skipping
ProjectStep[16477] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3b0f38d8> - skipping
ProjectStep[16478] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3a908970> - skipping
ProjectStep[16479] error migrating schedule parent: undefined method `schedule_parent=' for #<ProjectGroup:0x007fcf3b08e078> - skipping
```

**Comment by X0nic** [comment]
```sh
❯ cat migration_errors.log
ProjectEvent[402] - mapping 0 MemberId ref to null
ProjectEvent[403] - mapping 0 MemberId ref to null
ProjectEvent[404] - mapping 0 MemberId ref to null
ProjectEvent[405] - mapping 0 MemberId ref to null
ProjectEvent[406] - mapping 0 MemberId ref to null
ProjectEvent[407] - mapping 0 MemberId ref to null
ProjectEvent[408] - mapping 0 MemberId ref to null
ProjectEvent[409] - mapping 0 MemberId ref to null
ProjectEvent[462] - mapping 0 MemberId ref to null
ProjectEvent[465] - mapping 0 MemberId ref to null
ProjectEvent[4139] - mapping 0 MemberId ref to null
ProjectEvent[14316] - Person not found for MemberId: 249, mapping MemberId ref to null
ProjectGroup[16102] error migrating parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectGroup[16103] error migrating parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectGroup[16104] error migrating parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
ProjectGroup[16105] error migrating parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
ProjectStep[16113] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16114] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16115] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16116] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16342] error migrating parent: Couldn't find TimelineEntry with 'id'=16336 - skipping
ProjectGroup[15793] error migrating schedule parent: expected numeric - skipping
ProjectStep[15795] error migrating schedule parent: ArgumentError - skipping
ProjectStep[15797] error migrating schedule parent: expected numeric - skipping
ProjectGroup[16102] error migrating schedule parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectGroup[16104] error migrating schedule parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
```

**Comment by X0nic** [comment]
I decided to just go ahead and move the schedule logic into TimelineEntry. I think this is what we want.

File: spec/models/project_group_spec.rb [hunk 1]
```diff
@@ -55,4 +55,148 @@
       expect(group.descendant_step_count).to eq group.children.size
     end
   end
+
```
Comment (X0nic) [inline spec/models/project_group_spec.rb]: I have duplicated the project_step_spec logic here. If this is how we want to go forward, I can extract into a shared spec.

**Comment by hooverlunch** [comment]
What is causing the one `ArgumentError` in the second log file dump you shared?

**Comment by X0nic** [comment]
```
❯ cat migration_errors.log
ProjectEvent[402] - mapping 0 MemberId ref to null
ProjectEvent[403] - mapping 0 MemberId ref to null
ProjectEvent[404] - mapping 0 MemberId ref to null
ProjectEvent[405] - mapping 0 MemberId ref to null
ProjectEvent[406] - mapping 0 MemberId ref to null
ProjectEvent[407] - mapping 0 MemberId ref to null
ProjectEvent[408] - mapping 0 MemberId ref to null
ProjectEvent[409] - mapping 0 MemberId ref to null
ProjectEvent[462] - mapping 0 MemberId ref to null
ProjectEvent[465] - mapping 0 MemberId ref to null
ProjectEvent[628] - Loan 111 not found
ProjectEvent[629] - Loan 111 not found
ProjectEvent[630] - Loan 111 not found
ProjectEvent[741] - Loan 118 not found
ProjectEvent[823] - Loan 130 not found
ProjectEvent[1637] - Loan 219 not found
ProjectEvent[1709] - Loan 0 not found
ProjectEvent[1757] - Loan 224 not found
ProjectEvent[1768] - Loan 0 not found
ProjectEvent[1855] - Loan 0 not found
ProjectEvent[1998] - Loan 278 not found
ProjectEvent[2000] - Loan 279 not found
ProjectEvent[2001] - Loan 279 not found
ProjectEvent[2298] - Loan 324 not found
ProjectEvent[2299] - Loan 324 not found
ProjectEvent[2300] - Loan 324 not found
ProjectEvent[2301] - Loan 324 not found
ProjectEvent[2516] - Loan 348 not found
ProjectEvent[2608] - Loan 67 not found
ProjectEvent[2621] - Loan 278 not found
ProjectEvent[2853] - Loan 375 not found
ProjectEvent[3191] - Loan 426 not found
ProjectEvent[3192] - Loan 426 not found
ProjectEvent[3193] - Loan 426 not found
ProjectEvent[3194] - Loan 426 not found
ProjectEvent[3195] - Loan 426 not found
ProjectEvent[3196] - Loan 426 not found
ProjectEvent[3207] - Loan 429 not found
ProjectEvent[3214] - Loan 429 not found
ProjectEvent[3215] - Loan 429 not found
ProjectEvent[3216] - Loan 429 not found
ProjectEvent[3471] - Loan 442 not found
ProjectEvent[3472] - Loan 442 not found
ProjectEvent[3717] - Loan 476 not found
ProjectEvent[4034] - Loan 503 not found
ProjectEvent[4035] - Loan 503 not found
ProjectEvent[4069] - Loan 514 not found
ProjectEvent[4095] - Loan 514 not found
ProjectEvent[4096] - Loan 514 not found
ProjectEvent[4105] - Loan 514 not found
ProjectEvent[4494] - Loan 0 not found
ProjectEvent[4699] - Loan 594 not found
ProjectEvent[4700] - Loan 594 not found
ProjectEvent[5085] - Loan 641 not found
ProjectEvent[5086] - Loan 641 not found
ProjectEvent[5420] - Loan 677 not found
ProjectEvent[5629] - Loan 677 not found
ProjectEvent[5630] - Loan 677 not found
ProjectEvent[7851] - Loan 875 not found
ProjectEvent[9485] - Loan 991 not found
ProjectEvent[9486] - Loan 991 not found
ProjectEvent[9487] - Loan 991 not found
ProjectEvent[9488] - Loan 991 not found
ProjectEvent[9489] - Loan 991 not found
ProjectEvent[9490] - Loan 991 not found
ProjectEvent[9491] - Loan 991 not found
ProjectEvent[9492] - Loan 991 not found
ProjectEvent[9493] - Loan 991 not found
ProjectEvent[9960] - Loan 1022 not found
ProjectEvent[9961] - Loan 1022 not found
ProjectEvent[10733] - Loan 1081 not found
ProjectEvent[10734] - Loan 1081 not found
ProjectEvent[10735] - Loan 1081 not found
ProjectEvent[10736] - Loan 1081 not found
ProjectEvent[13800] - Loan 1319 not found
ProjectEvent[13801] - Loan 1319 not found
ProjectEvent[13802] - Loan 1319 not found
ProjectGroup[16102] error migrating parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectGroup[16103] error migrating parent: Couldn't find TimelineEntry with 'id'=16098 - skipping
ProjectGroup[16104] error migrating parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
ProjectGroup[16105] error migrating parent: Couldn't find TimelineEntry with 'id'=16099 - skipping
ProjectStep[16113] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16114] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16115] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16116] error migrating parent: Couldn't find TimelineEntry with 'id'=16101 - skipping
ProjectStep[16342] error migrating parent: Couldn't find TimelineEntry with 'id'=16336 - skipping
Couldn't find Legacy::ProjectEvent with 'ID'=16098
ProjectGroup[16102]: Missing Date, DependentDate[16098]
Couldn't find Legacy::ProjectEvent with 'ID'=16099
ProjectGroup[16104]: Missing Date, DependentDate[16099]
```

**Pull request merged**